content/blogs/gui-setup-part3-first-testing-problems.md

```diff
@@ -0,0 +1,267 @@
+1. Get the next file source/target paths from the queue;
+2. Call `do_file_copyW()` for the file: if it exists in the source installation directory, copy it and Goto Step 1.
+3. If not, check installation CAB file (`reactos.cab` in our case) with `extract_cabinet_file()`.
+If failed, throw error via the file-copy callback.
+4. Open CAB file, loop through it to find the file, extract it, and close the CAB file.
+5. Goto Step 1 for next file.
+
+In Steps 3. and 4., the helper function `extract_cabinet_file()` is used, and ultimately calls into the `cabinet.dll!FDICopy()` function (`cabinet.dll` is also a Wine-synced dll).
+Finding one given file in a CAB archive is a costly operation (with a ~80MB sized archive, if the file is at its end, finding it can take approximately 1 second), due to [how CAB files work](MSDN_link_for_CAB_files).
+
+These two steps are the bottleneck of this whole file-copy loop:
+because, for each file to copy, potentially present in the _same_ CAB archive, the archive is opened, the file has to be found (by looping through them to find the correct one) and extracted, then the CAB archive is closed.
+And this operation is done over and over again.
+
+In comparison, it is _much faster_ to open the CAB archive _only once_, then whenever a file needs to be copied, check whether it exists in the CAB archive, starting from the current search position, and if possible, exploit the fact that the files in the copy queue are sorted in order of their appearance in the CAB archive.
+(Alternatively, loop once through all the files in the CAB archive and check whether each successive of these are files that need to be extracted.)
+The exception could be, that if the next file is not in the order, one restarts finding from the beginning of the CAB archive.
+This would be the slow code path, but not actually taken in real life due to the reasons from above.
+
+This is exactly what the text-mode USETUP installer does, and why ReactOS installation time is there blazing fast in comparison.
+
+Additionally, it appears, after some further investigation (suggested by Mark Jansen), that each read operation from the CAB file, implemented with a `fdi->read()` callback (registered when initializing the fdi context with the `FDICreate()` function, done by setupapi), that does a `ReadFile()` call.
+Meaning, basically, that each data read from the CAB triggers a file I/O operation to disk, making the whole operation slower than it could be.
+This could be alleviated by adding ReactOS-specific code in setupapi, to memory-map the CAB archive instead so that reading from it can be done directly via memory reads.
+Such a thing is also done by the text-mode USETUP.
+
+#### The current workaround
+
+To alleviate the problems described above, the simplest workaround is to just use USETUP's file-copy and extraction functions. With minimal adaptations the existing code can be easily adapted for usage in the GUI installer. File copy works and finishes speedily.
```

Comment:
```suggestion
To alleviate the problems described above, the simplest workaround is to just use USETUP's file-copy and extraction functions. With minimal adaptations the existing code can be easily reused for usage in the GUI installer. File copy works and finishes quickly.
```
(also there are several sentences, please one per line)

Reply:
"reused for usage" --> "reused"